doc/api/stream.md

```diff
@@ -1159,6 +1159,31 @@ readable stream will release any internal resources.
 Implementors should not override this method, but instead implement
 [`readable._destroy`][readable-_destroy].
 
+##### readable[Symbol.asyncIterator]
+<!-- YAML
+added: REPLACEME
+-->
+
+> Stability: 1 - Experimental
+
+Returns an [AsyncIterator][async-iterator] to fully consume the stream.
+
+```js
+async function print (readable) {
+  readable.setEncoding('utf8');
+  let data = '';
+  for await (const k of readable) {
+    k += data;
+  }
+  console.log(data);
+}
+
+print(fs.createReadStream('file')).catch(console.log);
+```
+
+If you break or throw from within the for-await loop, the stream will be
```

Comment:
I mean:

```js
for await (const k of readable) {
  break;
}
```

this would call `return()` on the AsyncIterator which will call `.destroy()`.
This is debatable, as after `.destroy()` the readable cannot be used anymore.

However:

```js
for await (const k of readable) {
  throw new Error('kaboom');
}
```

would call `return()` __as well__ on the AsyncIterator which will call `.destroy()`.
This is the correct behavior, otherwise we would want to do all the time:

 ```js
try { 
  for await (const k of readable) {
    throw new Error('kaboom')
  }
} finally {
  readable.destroy()
}
```

Which will be prone to file descriptor leaking.

I'm not aware of a way to distinguish these two flows.

Should I document it in this way? Do you agree with the behavior?

Reply:
I agree with the behavior 100%, once you've started consuming an async iterator for a `for await` loop it should probably close if I break.

Pinging @zenparsing - I remember a lot of discussion about this - what should be the behavior in your opinion?


As for documentation - that is exactly what I meant - we should document that `destroy` will be called (rather than that the stream is destroyed) since it is a better guarantee for people subclassing ReadableStream.

Reply:
Calling `destroy` is consistent with how async generator functions behave after calling `return` on their iterators, so 👍 .